### PR TITLE
js: ApplyCustomUserTags

### PR DESCRIPTION
**Pull request by codebien**
Refactoring of the logic for applying user-defined tags in a centralized function.
It also includes the change for always raising an exception in case of a denied type is provided. https://github.com/grafana/k6/issues/2653

Closes https://github.com/grafana/k6/issues/2653

### CODE REVIEW

File: js/modules/k6/execution/execution.go
Comment (na--) [inline js/modules/k6/execution/execution.go:316]: This comment about Throw is invalid, see https://github.com/grafana/k6/issues/2653
I think this PR copied the parts of https://github.com/grafana/k6/pull/2654 that fixed the issue? Though I see that even in my PR I forgot to adjust the comment :sweat_smile:

File: lib/netext/httpext/request.go
Comment (na--) [reply lib/netext/httpext/request.go:184]: and btw the fact that no tests failed probably means we are lacking tests for this feature :disappointed:

File: js/modules/k6/grpc/client_test.go
Comment (codebien) [reply js/modules/k6/grpc/client_test.go:660]: I think I did a mistake in resolve a conflict in the original PR, going to fix

File: lib/netext/httpext/request.go
Comment (codebien) [reply lib/netext/httpext/request.go:184]: Good catch, going to add one

File: js/modules/k6/execution/execution.go
Comment (codebien) [reply js/modules/k6/execution/execution.go:316]: Yeah, I would split the fix in one more commit, going to do soon

File: js/modules/k6/ws/ws.go
Comment (olegbespalov) [inline js/modules/k6/ws/ws.go:160]: Not sure what the end user-faced error looks like, but if it's generic, I'd also suggest explicitly mentioning that this is coming from the WebSockets module. And basically it's relevant for any place like HTTP, groups, or whatever.

My point if we explicitly say "where" the issue is to the users, they will be happy :relaxed:

File: lib/netext/httpext/request_test.go
Comment (olegbespalov) [inline lib/netext/httpext/request_test.go:515]: Please also use `logger.Out = io.Discard`

File: js/common/tags.go
Comment (codebien) [reply js/common/tags.go:15]: Yeah good spot, I don't think it's quite relevant. The function sets a tag from a JavaScript value, I guess the **custom user** part would enforce that, but having them in the `js` package probably should be enough.

@na-- coined the name so proxying the question to him if it was a different reason.

File: js/common/tags.go
Comment (codebien) [reply js/common/tags.go:52]: Why do you want to do that? This will be hit by the users, I think we should keep a user-friendly experience. A code's function name reported in the error I don't helps. I'm happy to improve the message if you think is not clear yet enough.

Comment (olegbespalov) [reply js/common/tags.go:52]: Nope, I didn't mean that we should mention the function name in error what I meant was that the error message ideally should match the function name, which helps maintainers/contributors. 

Like since the error message is saying `"setting '%s' metric tag failed;"` it will be good if the method names something like `SetTag` instead of the applying. Or if the `ApplyTag` will be chosen then the error message should be `"...applying tag failed..."`. So the point is that they should be consistent.

Comment (codebien) [reply js/common/tags.go:52]: Gotcha, I agree! I wait for the final decision about the naming then I will change this in a consistent way.

File: js/modules/k6/execution/execution_test.go
Comment (codebien) [reply js/modules/k6/execution/execution_test.go:136]: No, it shouldn't and this is some not well-migrated part. I'm going to re-arrange it better.

File: js/common/tags.go
Comment (na--) [reply js/common/tags.go:15]: A bunch of things in the `js` package (almost exclusively, in fact :sweat_smile:) also emit the system metric tags, so I think it makes sense to keep the current name. After all, we mostly do a bunch of validation here, which we don't do for non-user-specified tags.

File: js/common/tags.go
Comment (na--) [reply js/common/tags.go:52]: Since users would usually see this error in another context, e.g. when making an http request with `http.get(url, tags: {foo: somethingInvalid})`, I'd suggest something like this:
```suggestion
			"invalid value for metric tag '%s': "+
```

File: js/modules/k6/execution/execution.go
Comment (na--) [inline js/modules/k6/execution/execution.go:314]: Please also add a mention in the PR description that it fixes https://github.com/grafana/k6/issues/2653

File: js/modules/k6/metrics/metrics.go
Comment (na--) [inline js/modules/k6/metrics/metrics.go:114]: this error wrapping is more confusing to the user than helpful... they don't care that we internally "apply" tags, they are just passing a JS object with `{key: value}` tags. So, there either shouldn't be any error wrapping here, or it should be something specific for where the error originates, like this:
```suggestion
		return false, fmt.Errorf("invalid custom metric tags: %w", err)
```
or maybe even add the `Metric`'s name in the error? :thinking:

File: js/modules/k6/http/request.go
Comment (na--) [inline js/modules/k6/http/request.go:313]: Same as https://github.com/grafana/k6/pull/2716/files#r990395132, "apply" shouldn't be part of the error because the users are not "applying" anything. Either no wrapping or something like this:
```suggestion
					return nil, fmt.Errorf("invalid HTTP request metric tags: %w", err)
```

File: js/modules/k6/ws/ws.go
Comment (na--) [inline js/modules/k6/ws/ws.go:160]: same as the other comments

**Review by na-- — DISMISSED** [review]
LGTM except the minor issues I left as inline comments

File: js/modules/k6/metrics/metrics.go
Comment (codebien) [reply js/modules/k6/metrics/metrics.go:114]:
```suggestion
		return false, fmt.Errorf("'%s' custom metric defines invalid tags: %w", m.metric.Name, err)
```
Changed with this

File: js/modules/k6/grpc/client.go
Comment (olegbespalov) [inline js/modules/k6/grpc/client.go:87]: Maybe be more precise, since Connect is a generic one? WDYT?
```suggestion
		return false, fmt.Errorf("grpc.Connect has invalid parameters: %w", err)
```

File: js/modules/k6/grpc/client.go
Comment (olegbespalov) [inline js/modules/k6/grpc/client.go:166]: Same
```suggestion
		return nil, fmt.Errorf("grpc.Invoke has invalid parameters: %w", err)
```

File: js/modules/k6/ws/ws.go
Comment (olegbespalov) [inline js/modules/k6/ws/ws.go:160]:
```suggestion
					return nil, fmt.Errorf("ws.Connect has metric tags: %w", err)
```

**Review by olegbespalov — DISMISSED** [review]
LGTM in general, left a few non-blocking comments :+1:

File: js/modules/k6/execution/execution.go
Comment (codebien) [inline js/modules/k6/execution/execution.go:352]: >The order is not deterministic.

This is not true anymore, going to remove

File: js/modules/k6/grpc/client.go
Comment (na--) [inline js/modules/k6/grpc/client.go:120]:
```suggestion
		return false, fmt.Errorf("invalid grpc.Connect() parameters: %w", err)
```

File: js/modules/k6/grpc/client.go
Comment (na--) [inline js/modules/k6/grpc/client.go:234]: Hmm where is this done now? he `name` tag should be set to be equal to the `url` tag, if the user didn't explicitly specify it.

File: js/modules/k6/metrics/metrics.go
Comment (na--) [inline js/modules/k6/metrics/metrics.go:114]:
```suggestion
		return false, fmt.Errorf("cannot add tags for the '%s' custom metric: %w", m.metric.Name, err)
```

**Comment by na--** [comment]
LGTM except the error message nitpicks and the gRPC `name` tag issue.

File: js/modules/k6/grpc/client.go
Comment (codebien) [reply js/modules/k6/grpc/client.go:234]: mmm...good catch, my fault. And apparently, we are not covering with tests this part.

**Comment by codebien** [comment]
@olegbespalov do you agree with the latest suggestions?

File: js/modules/k6/grpc/client.go
Comment (na--) [reply js/modules/k6/grpc/client.go:234]: :disappointed: yeah...

**Comment by olegbespalov** [comment]
yeah, the most important they're pointing to exact place, so I'm happy :relaxed:

File: js/modules/k6/grpc/client.go
Comment (codebien) [reply js/modules/k6/grpc/client.go:234]: Adding the test I hit the issue that the URL is not set equal to the Name and I guess that is what we expect after https://github.com/grafana/k6/pull/2703 so I would prefer to add it in a dedicated bugfix PR.

**Review by olegbespalov — APPROVED** [review]
LGTM, with one nit (non-blocking) :+1:

File: js/modules/k6/execution/execution.go
Comment (olegbespalov) [reply js/modules/k6/execution/execution.go:320]: I see :+1: anyway it's was just a non-blocking nit :sweat_smile:

File: js/modules/k6/grpc/client.go
Comment (na--) [reply js/modules/k6/grpc/client.go:234]: Hmm IIRC the `url` tag in gRPC has a much lower cardinality than in HTTP requests, it is just the `server/method_name`, right? So I am not sure we need to have the same "`url` tag should always be equal to the `name` tag" logic in the gRPC module that we had to add to the `k6/http` module. gRPC doesn't have any GET parameters, which is the big problem with HTTP.

File: js/modules/k6/grpc/client.go
Comment (na--) [inline js/modules/k6/grpc/client.go:120]:
```suggestion
		return false, fmt.Errorf("invalid grpc.connect() parameters: %w", err)
```

The actual JS method name is with lowercase, see https://k6.io/docs/javascript-api/k6-net-grpc/client/client-connect

File: js/modules/k6/grpc/client.go
Comment (na--) [inline js/modules/k6/grpc/client.go:199]:
```suggestion
		return nil, fmt.Errorf("invalid grpc.invoke() parameters: %w", err)
```

https://k6.io/docs/javascript-api/k6-net-grpc/client/client-invoke

File: js/modules/k6/ws/ws.go
Comment (na--) [inline js/modules/k6/ws/ws.go:160]:
```suggestion
					return nil, fmt.Errorf("invalid ws.connect() metric tags: %w", err)
```

https://k6.io/docs/javascript-api/k6-ws/connect/

**Comment by na--** [comment]
LGTM except the wrong method capitalization in error messages

File: js/modules/k6/grpc/client.go
Comment (codebien) [reply js/modules/k6/grpc/client.go:234]: Added an issue for addressing it with a specific discussion in a issue + PR. https://github.com/grafana/k6/issues/2722

**Review by na-- — APPROVED** [review]

**Review by olegbespalov — APPROVED** [review]

**Pull request merged**
